#include "DeployMaster.h"
#include "TelnetClient.h"

#include <QFileInfo>
#include <QFileDialog>
#include <QString>
#include <QtConcurrent/QtConcurrent>
#include <QStyleFactory>

DeployMaster::DeployMaster(QWidget* parent)
    : QMainWindow(parent)
{
    ui.setupUi(this);

    QApplication::setStyle(QStyleFactory::create("Fusion"));

    connect(ui.btn_addFiles, &QPushButton::clicked, this, &DeployMaster::onAddFilesClicked);
    connect(ui.btn_addFolders, &QPushButton::clicked, this, &DeployMaster::onAddFolderClicked);
    connect(ui.btn_clearUploadList, &QPushButton::clicked, this, &DeployMaster::onFileItemCleanClicked);
    connect(ui.list_uploadedItems, &DropListWidget::filesDropped, this, &DeployMaster::onFilesDropped);

    connect(ui.btn_deploy, &QPushButton::clicked, this, &DeployMaster::onDeployClicked);
    connect(ui.btn_clearLog, &QPushButton::clicked, this, &DeployMaster::onClearLogClicked);
}

void DeployMaster::onAddFilesClicked()
{
    // è®¾ç½®æ–‡ä»¶å¯¹è¯æ¡†æ ‡é¢˜å’Œè¿‡æ»¤å™¨
    QString title = tr("é€‰æ‹©è¦éƒ¨ç½²çš„æ–‡ä»¶");
    QString filter = tr("æ‰€æœ‰æ–‡ä»¶ (*.*);;å¯æ‰§è¡Œæ–‡ä»¶ (*.exe);;é…ç½®æ–‡ä»¶ (*.ini *.conf);;è„šæœ¬æ–‡ä»¶ (*.py *.sh)");

    // å¼¹å‡ºæ–‡ä»¶é€‰æ‹©å¯¹è¯æ¡†ï¼Œå…è®¸å¤šé€‰
    QStringList selectedFiles = QFileDialog::getOpenFileNames(this, title, lastUsedDirectory, filter);

    if (!selectedFiles.isEmpty()) {
        // æ›´æ–°â€œä¸Šæ¬¡ä½¿ç”¨çš„ç›®å½•â€ï¼Œæ–¹ä¾¿ä¸‹æ¬¡æ‰“å¼€
        lastUsedDirectory = QFileInfo(selectedFiles.first()).absolutePath();

        // å°†é€‰ä¸­çš„æ–‡ä»¶æ·»åŠ åˆ°ä½ çš„å†…éƒ¨æ•°æ®æ¨¡å‹ï¼ˆä¾‹å¦‚ QStringList, QList<QFileInfo>, æˆ–è‡ªå®šä¹‰ Modelï¼‰
        for (const QString& filePath : selectedFiles) {
            addFileToList(filePath); // ä½ éœ€è¦å®ç°è¿™ä¸ªå‡½æ•°ï¼Œå°†æ–‡ä»¶è·¯å¾„æ·»åŠ åˆ° QListWidget/QTableView ç­‰æ§ä»¶ä¸­
        }
    }
}

void DeployMaster::onAddFolderClicked()
{
    QString title = tr("é€‰æ‹©è¦éƒ¨ç½²çš„æ–‡ä»¶å¤¹");
    // å¼¹å‡ºæ–‡ä»¶å¤¹é€‰æ‹©å¯¹è¯æ¡†
    QString selectedFolder = QFileDialog::getExistingDirectory(this, title, lastUsedDirectory);

    if (!selectedFolder.isEmpty()) {
        lastUsedDirectory = selectedFolder; // æ›´æ–°ä¸Šæ¬¡ç›®å½•

        // å°†é€‰ä¸­çš„æ–‡ä»¶å¤¹æ·»åŠ åˆ°åˆ—è¡¨
        addFolderToList(selectedFolder); // ä½ éœ€è¦å®ç°è¿™ä¸ªå‡½æ•°
    }
}

void DeployMaster::onDeployClicked()
{
    QString user = ui.txt_user->text().trimmed();
    QString pass = ui.txt_pass->text().trimmed(); // æ³¨æ„ï¼šQt æ—  PasswordBoxï¼Œéœ€ç”¨ QLineEdit + setEchoMode
    bool shouldReboot = ui.chk_rebootAfterDeploy->isChecked();
    bool shouldClearRemote = ui.chk_clearRemoteBeforeUpload->isChecked();
    QString remotePath = ui.txt_remotePath->text().trimmed();
    if (!remotePath.endsWith('/')) remotePath += '/';

    if (uploadItems.isEmpty()) {
        appendFtpLog("âŒ é”™è¯¯ï¼šè¯·è‡³å°‘æ·»åŠ ä¸€ä¸ªè¦ä¸Šä¼ çš„æ–‡ä»¶æˆ–æ–‡ä»¶å¤¹ï¼");
        return;
    }
    if (user.isEmpty() || pass.isEmpty()) {
        appendFtpLog("âŒ é”™è¯¯ï¼šè¯·è¾“å…¥ç”¨æˆ·åå’Œå¯†ç ï¼");
        return;
    }

    QStringList ips = getTargetIPs();
    if (ips.isEmpty()) {
        appendFtpLog("âŒ é”™è¯¯ï¼šè¯·è‡³å°‘è¾“å…¥ä¸€ä¸ªç›®æ ‡ IPï¼");
        return;
    }

    ui.btn_deploy->setEnabled(false);
    appendFtpLog(QString("ğŸš€ å¼€å§‹éƒ¨ç½² %1 ä¸ªå†…å®¹åˆ° %2 å°è®¾å¤‡...")
        .arg(uploadItems.size()).arg(ips.size()));

    // å¯åŠ¨å¼‚æ­¥ FTP ä¸Šä¼ ï¼ˆä»…ä¸Šä¼ ï¼Œä¸å¤„ç† Telnetï¼‰
    QtConcurrent::run([=]() {
        QStringList deploySuccesses, deployFailures;

        for (const QString& ip : ips) {
            QString targetIp = ip.trimmed();
            QString port = targetIp.contains(':') ? "" : "21";

            FtpManager ftpm(targetIp, port.toUShort(), user, pass);
            // æ³¨æ„ï¼šappendFtpLog ä¸æ˜¯çº¿ç¨‹å®‰å…¨çš„ï¼éœ€ç”¨ invokeMethod
            QMetaObject::invokeMethod(this, "appendFtpLog", Qt::QueuedConnection,
                Q_ARG(QString, QString("â¡ï¸ æ­£åœ¨éƒ¨ç½²åˆ°è®¾å¤‡ï¼š%1").arg(targetIp)));

            // ğŸ”¥ æ–°å¢ï¼šæ¸…ç©ºè¿œç¨‹è·¯å¾„ï¼ˆå¦‚æœå‹¾é€‰ï¼‰
            if (shouldClearRemote) {
                try {
                    QMetaObject::invokeMethod(this, "appendFtpLog", Qt::QueuedConnection,
                        Q_ARG(QString, QString("ğŸ§¹ æ­£åœ¨æ¸…ç©ºè¿œç¨‹è·¯å¾„: %1").arg(remotePath)));
                    ftpm.clearRemoteDirectory(remotePath);
                    QMetaObject::invokeMethod(this, "appendFtpLog", Qt::QueuedConnection,
                        Q_ARG(QString, "âœ… è¿œç¨‹è·¯å¾„å·²æ¸…ç©º"));
                }
                catch (const std::exception& ex) {
                    QMetaObject::invokeMethod(this, "appendFtpLog", Qt::QueuedConnection,
                        Q_ARG(QString, QString("âŒ æ¸…ç©ºå¤±è´¥: %1").arg(QString::fromStdString(ex.what()).left(100))));
                    // å¯é€‰æ‹©ï¼šè·³è¿‡è¯¥è®¾å¤‡ï¼Œæˆ–ç»ˆæ­¢éƒ¨ç½²ï¼Ÿ
                    // è¿™é‡Œå»ºè®®è·³è¿‡ï¼Œè®°å½•å¤±è´¥
                    deployFailures << targetIp;
                    continue; // è·³è¿‡ä¸Šä¼ 
                }
            }

            bool allSuccess = true;
            for (const UploadItem& item : uploadItems) {
                try {
                    if (item.isFolder) {
                        ftpm.uploadFolder(item.fullPath, remotePath);
                    }
                    else {
                        ftpm.uploadFile(item.fullPath, remotePath);
                    }
                    QMetaObject::invokeMethod(this, "appendFtpLog", Qt::QueuedConnection,
                        Q_ARG(QString, QString("âœ… ä¸Šä¼ æˆåŠŸ: %1").arg(item.displayName())));
                }
                catch (const std::exception& ex) {
                    QMetaObject::invokeMethod(this, "appendFtpLog", Qt::QueuedConnection,
                        Q_ARG(QString, QString("âŒ ä¸Šä¼ å¤±è´¥: %1 -> %2").arg(item.displayName(), QString::fromStdString(ex.what()).left(100))));
                    allSuccess = false;
                }
                QThread::msleep(50);
            }

            if (allSuccess) {
                deploySuccesses << targetIp;
            }
            else {
                deployFailures << targetIp;
            }
            QThread::msleep(100);
        }
        QMetaObject::invokeMethod(this, "appendFtpLog", Qt::QueuedConnection,
            Q_ARG(QString, QString("ğŸ‰ æ‰¹é‡éƒ¨ç½²å®Œæˆï¼")));
        QMetaObject::invokeMethod(this, "appendFtpLog", Qt::QueuedConnection,
            Q_ARG(QString, QString("ğŸ“Š éƒ¨ç½²æˆåŠŸ: %1 | éƒ¨ç½²å¤±è´¥: %2").arg(deploySuccesses.size()).arg(deployFailures.size())));
        if (deployFailures.size() > 0) {
            QMetaObject::invokeMethod(this, "appendFtpLog", Qt::QueuedConnection,
                Q_ARG(QString, QString("âŒ å¤±è´¥åˆ—è¡¨:").arg(deployFailures.join(", "))));
        }
        // ä¸Šä¼ å®Œæˆåï¼Œå›åˆ°ä¸»çº¿ç¨‹ï¼šå…ˆæ±‡æŠ¥ç»“æœï¼Œå†å†³å®šæ˜¯å¦é‡å¯
        QMetaObject::invokeMethod(this, "onFtpUploadFinished", Qt::QueuedConnection,
            Q_ARG(QStringList, deploySuccesses),
            Q_ARG(QStringList, deployFailures),
            Q_ARG(bool, shouldReboot),
            Q_ARG(QString, user),
            Q_ARG(QString, pass)
        );
        });
}

void DeployMaster::onFtpUploadFinished(const QStringList& deploySuccesses,
    const QStringList& deployFailures,
    bool shouldReboot,
    const QString& user,
    const QString& pass)
{
    QStringList rebootSuccess, rebootFailure;
    if (shouldReboot && !deploySuccesses.isEmpty()) {
        appendFtpLog("ğŸ”„ å¼€å§‹å‘é€é‡å¯å‘½ä»¤...");

        // ä¸ºæ¯ä¸ªæˆåŠŸè®¾å¤‡åˆ›å»º TelnetClientï¼ˆåœ¨ä¸»çº¿ç¨‹ï¼ï¼‰
        for (const QString& ip : deploySuccesses) {
            TelnetClient tc;

            if (!tc.syncConnectToHost(ip, 23, 3000)) {
                rebootFailure << ip;
                appendFtpLog(QString("âŒ è¿æ¥å¤±è´¥: %1").arg(ip));
                continue;
            }

            if (!tc.syncLogin(user, pass, 5000)) {
                rebootFailure << ip;
                appendFtpLog(QString("âŒ ç™»å½•å¤±è´¥: %1").arg(ip));
                continue;
            }

            if (tc.syncSendCommand("reboot", 1000)) {
                rebootSuccess << ip;
                appendFtpLog(QString("âœ… é‡å¯å‘½ä»¤å·²å‘é€: %1").arg(ip));
            }
            else {
                rebootFailure << ip;
                appendFtpLog(QString("âš ï¸ å‘½ä»¤å‘é€å¼‚å¸¸: %1").arg(ip));
            }
        }

        appendFtpLog("ğŸ‰ æ‰¹é‡é‡å¯å®Œæˆï¼");
        appendFtpLog(QString("ğŸ“Š é‡å¯æˆåŠŸ: %1 | é‡å¯å¤±è´¥: %2")
            .arg(rebootSuccess.size()).arg(rebootFailure.size()));
        if (!rebootFailure.isEmpty())
            appendFtpLog("âŒ å¤±è´¥åˆ—è¡¨: " + rebootFailure.join(", "));
    }

    ui.btn_deploy->setEnabled(true);
}

void DeployMaster::onFilesDropped(const QStringList& filePaths)
{
    for (const QString& path : filePaths) {
        QFileInfo fileInfo(path);
        if (fileInfo.isFile()) {
            addFileToList(path); // å¤ç”¨ä¹‹å‰çš„æ·»åŠ æ–‡ä»¶é€»è¾‘
        }
        else if (fileInfo.isDir()) {
            addFolderToList(path); // å¤ç”¨ä¹‹å‰çš„æ·»åŠ æ–‡ä»¶å¤¹é€»è¾‘
        }
        // æ³¨æ„ï¼šæ‹–æ‹½å¯ä»¥åŒæ—¶åŒ…å«æ–‡ä»¶å’Œæ–‡ä»¶å¤¹ï¼Œæ‰€ä»¥éœ€è¦åˆ¤æ–­ç±»å‹
    }
}

void DeployMaster::onFileItemCleanClicked()
{
    ui.list_uploadedItems->clear();
    uploadItems.clear();
}

void DeployMaster::onClearLogClicked()
{
    ui.txt_globalLog->clear();
}

void DeployMaster::addFileToList(const QString& filePath)
{
    // æ£€æŸ¥æ–‡ä»¶æ˜¯å¦å·²ç»åœ¨åˆ—è¡¨ä¸­ï¼Œé¿å…é‡å¤æ·»åŠ 
    QList<QListWidgetItem*> existingItems = ui.list_uploadedItems->findItems(filePath, Qt::MatchExactly);
    if (existingItems.isEmpty()) {
        ui.list_uploadedItems->addItem(filePath);
        UploadItem item;
        item.isFolder = false;
        item.fullPath = filePath;
        uploadItems.push_back(item);
    }
}

void DeployMaster::addFolderToList(const QString& folderPath)
{
    // æ£€æŸ¥æ–‡ä»¶å¤¹æ˜¯å¦å·²ç»åœ¨åˆ—è¡¨ä¸­ï¼Œé¿å…é‡å¤æ·»åŠ 
    QList<QListWidgetItem*> existingItems = ui.list_uploadedItems->findItems(folderPath, Qt::MatchExactly);
    if (existingItems.isEmpty()) {
        ui.list_uploadedItems->addItem(folderPath);
        UploadItem item;
        item.isFolder = true;;
        item.fullPath = folderPath;
        uploadItems.push_back(item);
    }
}

void DeployMaster::appendFtpLog(const QString& log)
{
    ui.txt_globalLog->append(log);
}

QStringList DeployMaster::getTargetIPs()
{
    QStringList ips;
    QString text = ui.txt_ipList->toPlainText(); // é€‚ç”¨äº QTextEdit å’Œ QPlainTextEdit

    // æŒ‰è¡Œåˆ†å‰²
    QStringList lines = text.split('\n', Qt::SkipEmptyParts);

    for (QString line : lines) {
        line = line.trimmed(); // å»é™¤é¦–å°¾ç©ºæ ¼
        if (!line.isEmpty()) {
            // å¯é€‰ï¼šç®€å•éªŒè¯æ˜¯å¦åƒ IPï¼ˆå¢å¼ºå¥å£®æ€§ï¼‰
            // ä¾‹å¦‚ï¼šè·³è¿‡æ˜æ˜¾æ— æ•ˆçš„è¡Œï¼ˆå¦‚åŒ…å«å­—æ¯ã€å¤šä¸ªå†’å·ç­‰ï¼‰
            // è¿™é‡Œå…ˆåªåšåŸºæœ¬è¿‡æ»¤
            ips << line;
        }
    }

    return ips;
}

DeployMaster::~DeployMaster()
{
}